### PR TITLE
[APP-15115] - Add integration-sdk-entity-validator package

### PR DESCRIPTION
**Pull request by geovannimp**

### CODE REVIEW

File: packages/integration-sdk-entity-validator/src/__tests__/validator.test.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  },
+  required: ['_key', '_class', '_type'],
+};
+
+describe('validator', () => {
+  let entityValidator: EntityValidator;
+  let classSchemas: AnySchema[] = [];
+
+  beforeAll(async () => {
+    classSchemas = await fetch(RESOLVED_SCHEMAS_URL)
```
Comment (geovannimp) [inline packages/integration-sdk-entity-validator/src/__tests__/validator.test.ts:53]: The idea is to export a function in the dataModel that does this and returns the latest data model version.

File: packages/integration-sdk-entity-validator/src/__tests__/validator.test.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      errors: null,
+      validationType: 'type',
+    });
+  });
+
+  test('Should validate with type if both entity type and class schemas are loaded', async () => {
+    const graphObjectSchema = classSchemas.find(
+      (s) => typeof s === 'object' && s.$id === '#GraphObject',
+    );
+    entityValidator.addSchemas(graphObjectSchema!);
```
Comment (geovannimp) [inline packages/integration-sdk-entity-validator/src/__tests__/validator.test.ts:107]: Just loading a single schema cause we have some duplicated required keys that ajv is complaining. As well is kinda a nice idea loading just the schemas that will be used, cause make it faster.

PR to fix duplicated required keys: https://github.com/JupiterOne/data-model/pull/200/files

File: packages/integration-sdk-entity-validator/src/validator.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import Ajv, { AnySchema, ErrorObject } from 'ajv';
+import addFormats from 'ajv-formats';
+import { addJ1Formats } from './j1Formats';
+
+export class EntityValidator {
+  private ajvInstance: Ajv;
+
+  constructor({ schemas }: { schemas?: AnySchema[] }) {
+    this.ajvInstance = addJ1Formats(
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/validator.ts:9]: I have a proposal for a different interface of this package, where this package provides the following functions:
- Fetch the _class schemas, e.g. `async fetchClassSchemas(): JsonSchemas[]`
- Fetch _type schemas of a certain integration, e.g. `async fetchTypeSchemas(integrationName: string): JsonSchemas[]`
- Run validation on an object (not `Entity`), with a passed in Ajv instance e.g.
```typescript
validateEntity(ajv: Ajv, obj: object): ValidationResult
```

This package **would not** have its own Ajv instance. That would instead be handled by the consumer, which allows them to configure the instance, and also handle its lifecycle (re-use it, or create a new one if they want).

What I'm proposing here does leak the detail of using Ajv to the consumer heavily, so you could say we're actually making an "Ajv utils" library not a generic validation library, but I think that's fine.

The reason for this drastic change of the API is that I don't think the validation logic can be abstracted away cleanly and promise the consumer that they don't need to know about how we actually run the validation.
Fetching the schemas is something the consumer will need to know about, especially fetching _type schemas, they might want to cache that or manage it in whatever way makes sense for them.
In addition, the results of the Ajv validator are not pleasant to work with. If we want to expose the Ajv validation result to the consumer, then we haven't really done anything useful to the consumer. They might as well just use Ajv by themselves without the need for this library.

**Review by ryan-willis — APPROVED** [review]

File: packages/integration-sdk-entity-validator/src/validator.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      }
+    }
+
+    return {
+      isValid: errors.length === 0,
+      errors: errors.length
+        ? errors.map(ajvErrorToEntityValidationError)
+        : null,
+      type: typeValidator ? 'type' : 'class',
+    };
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/validator.ts:67]: I like this. It's a good summary of what the function did.

File: packages/integration-sdk-entity-validator/src/validator.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      if (!isValid) {
+        errors = [...errors, ...(typeValidator.errors ?? [])];
+      }
+    } else {
+      for (const className of classArray) {
+        const validator = this.ajvInstance.getSchema(`#${className}`);
+
+        if (!validator) {
+          throw new Error(`Schema not found for class "${className}"`);
+        }
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/validator.ts:51]: ...which is why I don't like this. Throwing errors from inside a for loop is too "aggressive".
Can't we also have this information returned?
If we don't find a _class it makes sense to me that it would just be skipped, instead of halting the entire validation.

File: packages/integration-sdk-entity-validator/src/validator.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  addSchemas(schema: AnySchema | AnySchema[]) {
+    this.ajvInstance.addSchema(schema);
+    return this;
+  }
+
+  validateEntity<Entity extends { _class: string[] | string; _type: string }>(
+    entity: Entity,
+  ) {
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/validator.ts:31]: Do we want to change the parameter type to not expect an `Entity` type?

File: packages/integration-sdk-entity-validator/src/entityValidationError.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  message: string;
+  validation: string;
+};
+
+const getPropertyNameFromAjvError = (
+  error: ErrorObject,
+): EntityValidationError['property'] => {
+  // instancePath is either '/{propertyName}', or if it's a list: '/{propertyName}/index'
+  // instancePath will always start with a slash
+  return error.instancePath.split('/')[1] || unknownPropertySymbol;
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/entityValidationError.ts:16]: "instancePath will always start with a slash" does that also apply even if the error is an object-level error?

I think it does, but I can't be certain.

Because if there's ever an error where `instancePath == ""`, grabbing the index 1 of the split() will throw an error.

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 4 above)
Comment (eyadmba) [reply packages/integration-sdk-entity-validator/src/validator.ts:67]: `ajvErrorToEntityValidationError()` is much cleaner than what I had. Nice work.

File: packages/integration-sdk-entity-validator/src/entityValidationError.ts (hunk 7 above)
Comment (geovannimp) [reply packages/integration-sdk-entity-validator/src/entityValidationError.ts:16]: JS don't throw exceptions to "invalid" index because all arrays have "Infinite" size.
image

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 5 above)
Comment (geovannimp) [reply packages/integration-sdk-entity-validator/src/validator.ts:51]: That a really good point. I'm going to return a list of validated and not validated classes and let the consumer handle if it should throw or not the exception.

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 6 above)
Comment (geovannimp) [reply packages/integration-sdk-entity-validator/src/validator.ts:31]: I forgot about that. Let me fix it.

File: packages/integration-sdk-entity-validator/src/validator.ts [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    }
+  }
+
+  addSchemas(schema: AnySchema | AnySchema[]) {
+    this.ajvInstance.addSchema(schema);
+    return this;
+  }
+
+  validateEntity(
+    entity: object,
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/validator.ts:76]: How about we keep all the runtime checking you have, but change the type of the `entity` param to be `{_class: string; _type: string}` to have typescript help the consumer do the right thing?

File: packages/integration-sdk-entity-validator/src/validator.ts [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        }
+      } else {
+        skippedSchemas.push({
+          schemaId: typeSchemaId,
+          reason: 'not-found',
+          type: 'type',
+        });
+      }
+
+      if (!typeValidator || forceClassValidationWithValidatedType) {
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/validator.ts:112]: what's the use-case for `forceClassValidationWithValidatedType`?

File: packages/integration-sdk-entity-validator/src/validator.ts [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    entity: object,
+    {
+      forceClassValidationWithValidatedType = false,
+    }: ValidateEntityOptions = {},
+  ): ValidateEntityResult {
+    const errors: EntityValidationError[] = [];
+    const skippedSchemas: SkippedSchema[] = [];
+
+    try {
+      assertEntity(entity);
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/validator.ts:85]: I don't know if it's possible, but I'll say it anyway and leave it to you to decide. The nesting is making the code hard to follow, so if possible, it would be better to flatten out this function a little bit.

File: packages/integration-sdk-entity-validator/src/validator.ts [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        );
+      }
+    } catch (error) {
+      errors.push(convertUnknownErrorToEntityValidationError(error));
+    }
+
+    return {
+      isValid: errors.length === 0,
+      errors: errors.length ? errors : null,
+      skippedSchemas: skippedSchemas.length ? skippedSchemas : null,
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/validator.ts:155]: We lost the information of if we used the `_type` vs `_class` for the validation. Is that deliberate?

**Review by eyadmba — APPROVED** [review]
I left some comments and questions, but overall it looks good to me.

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 9 above)
Comment (geovannimp) [reply packages/integration-sdk-entity-validator/src/validator.ts:112]: I'm not sure, that because it's false by default. But you may want to validate the latest class schema as well cause the type schema has a outdated class schema merged in it.

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 11 above)
Comment (geovannimp) [reply packages/integration-sdk-entity-validator/src/validator.ts:155]: With the current implementation you could have both validations. You can get this information now from the skippedSchemas.

 Here is an example: 
image

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 8 above)
Comment (geovannimp) [reply packages/integration-sdk-entity-validator/src/validator.ts:76]: I'm not sure the type is useful. It can make people force the type to use the functions, like:

```typescript
validator.validateEntity(myEntityObject as unknown as {_class: string; _type: string})
```

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 9 above)
Comment (eyadmba) [reply packages/integration-sdk-entity-validator/src/validator.ts:112]: I was under the impression that when a class schema is updated, the type schemas that used it would also be updated (and published). But I also don't really know how that would work tbh.

File: packages/integration-sdk-entity-validator/src/validator.ts [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        skippedSchemas.push(
+          ...classArray.map(
+            (className) =>
+              ({
+                schemaId: `#${className}`,
+                reason: 'type-already-validated',
+                type: 'class',
+              }) satisfies SkippedSchema,
+          ),
+        );
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/validator.ts:146]: What are we doing here?

Comment (geovannimp) [reply packages/integration-sdk-entity-validator/src/validator.ts]: As we validated the type, we are adding all class schemas as skipped.

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 10 above)
Comment (geovannimp) [reply packages/integration-sdk-entity-validator/src/validator.ts:85]: Refactored. Everything is inside a single for loop right now making the code clear.

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 9 above)
Comment (geovannimp) [reply packages/integration-sdk-entity-validator/src/validator.ts]: That's the idea, but I'm not sure when we are getting there. So I think having this flag available might be helpful for now. In the best scenario it won't be used and we can remove it.

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 10 above)
Comment (eyadmba) [reply packages/integration-sdk-entity-validator/src/validator.ts:85]: It's beautiful!

File: packages/integration-sdk-entity-validator/src/validator.ts [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+            ...(validator.errors?.map((error) =>
+              ajvErrorToEntityValidationError(schemaId, error),
+            ) ?? []),
+          );
+        }
+
+        if (type === 'type' && !forceClassValidationWithValidatedType) {
+          skippedSchemas.push(
+            // Skip each other schema as the type should already have class properties injected
+            ...schemasToValidate.slice(Number(i) + 1).map(
```
Comment (eyadmba) [inline packages/integration-sdk-entity-validator/src/validator.ts:131]: This type of thing scares me. It's easy to miss when making changes, and it's tricky to test reliably.

Is there a way we can ensure that the schemasToValidate only contain the actual schemas we want to run. Like if we move that `forceClassValidationWithValidatedType` check up to when we build the `schemasToValidate` list?

File: packages/integration-sdk-entity-validator/src/validator.ts (hunk 8 above)
Comment (eyadmba) [reply packages/integration-sdk-entity-validator/src/validator.ts:76]: Doesn't that still indicate to the consumer that they're doing the wrong thing though? The type is doing its job still by guiding the consumer to not mess up unintentionally.

If they do want to over-rule the type error, they deserve all the evil things that will happen to them. In this life and the next.

**Review by eyadmba — APPROVED** [review]

File: packages/integration-sdk-entity-validator/src/j1Formats.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
+import Ajv, { Format as AjvFormat } from 'ajv';
+import addFormats from 'ajv-formats';
+
+const ipv4 = addFormats.get('ipv4') as RegExp;
+const ipv6 = addFormats.get('ipv6') as RegExp;
+
+const ipv4CidrRegex =
+  /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/(3[0-2]|[12]?[0-9])$/;
+const ipv6CidrRegex =
+  /^([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}\/(12[0-8]|1[01][0-9]|[1-9]?[0-9])$|^([0-9a-fA-F]{1,4}:){1,7}:\/(12[0-8]|1[01][0-9]|[1-9]?[0-9])$|^::\/(12[0-8]|1[01][0-9]|[1-9]?[0-9])$|^([0-9a-fA-F]{1,4}:){1,7}[0-9a-fA-F]{1,4}$/;
```
Comment (zemberdotnet) [inline packages/integration-sdk-entity-validator/src/j1Formats.ts:10]: Were these not exported from the data model package?

Comment (geovannimp) [reply packages/integration-sdk-entity-validator/src/j1Formats.ts:10]: Yep, but we'll be removing this (and ajv validation) from the data-model and using this package instead.
